fd.go

```diff
@@ -123,47 +134,62 @@ func (fi *fileDescriptor) Flush() error {
 
 // flushUp syncs the file and adds it to the dagservice
 // it *must* be called with the File's lock taken
-// TODO: What is `fullsync`? Propagate the changes upward
-// to the root flushing every node in the path (the "up"
-// part of `flushUp`).
+// If `fullSync` is set the changes are propagated upwards
+// (the `Up` part of `flushUp`).
 func (fi *fileDescriptor) flushUp(fullSync bool) error {
-	nd, err := fi.mod.GetNode()
-	if err != nil {
-		return err
-	}
+	var nd ipld.Node
+	switch fi.state {
+	case stateDirty:
+		// calls mod.Sync internally.
+		var err error
+		nd, err = fi.mod.GetNode()
+		if err != nil {
+			return err
+		}
+		err = fi.inode.dagService.Add(context.TODO(), nd)
+		if err != nil {
+			return err
+		}
+		fi.inode.nodeLock.Lock()
+		fi.inode.node = nd
+		fi.inode.nodeLock.Unlock()
+		fallthrough
+	case stateCreated:
+		if !fullSync {
```

Comment:
Two things are happening in this `case`:

1. The `File` is being updated with the new content (saving that information in the internal `inode.node`).
2. We let the parent know the file (and hence its hash/CID) has changed so it updates the corresponding entry (`updateChildEntry`).

(1) should always happen otherwise we lose the new contents, (2) needs to happen only if `fullSync` is set (we can always update the parent later); this *could* be part of the issue you're seeing, since if we don't update the internal node we are always left with an empty UnixFS file of size zero, that being said, the test seems to open the files it writes to with `fullSync` set so even if this check is wrong it should be bypassed in that particular test, either way we need to fix this.

Reply:
> (1) should always happen otherwise we lose the new contents, (2) needs to happen only if `fullSync` is set (we can always update the parent later); this _could_ be part of the issue you're seeing, since if we don't update the internal node we are always left with an empty UnixFS file of size zero, that being said, the test seems to open the files it writes to with `fullSync` set so even if this check is wrong it should be bypassed in that particular test, either way we need to fix this.

Commit https://github.com/ipfs/go-mfs/pull/49/commits/e2a0d399fbc7844ba4f1a1122a87eeb16d3f0825 should handle this.

Reply:
Sorry, I wasn't very clear I think, what I meant originally was just to move the `fullSync` to the `updateChildEntry` call after all the internal node manipulation, with the current patch we have now two distinct lock sections both accessing the same internal node (one writes it and one reads it but logically it's the same node, or it should be at least), we should:

1. Inside lock: update internal node and keep a reference, get parent, name (anything needed for the parent update).
2. If `fullSync` call `updateChildEntry`.

Reply:
Will do in the next commit. Just to make sure we still want to update the `inode` in both `stateDirty` and `stateCreated` right ? In that all I do is wrap `updateChildEntry` with the `fullSync` check.

Reply:
Yes.

So, from what I'm understanding now `stateCreated` and `stateDirty` share the same logic, shouldn't we (just asking, I'm not sure) just use `stateDirty` then and set new files with that state? Why did we have `stateCreated` in the first place then? (I'm a bit lost in the PR now and I may be missing something here).

Reply:
I suppose you are right for this case, but if we needed some special handling for fresh fd's versus modified fd's in the future I think it's a better idea to have two states. WDYT ?

Reply:
Created commit https://github.com/ipfs/go-mfs/pull/49/commits/ab71ad6b060b095ba3e16b5a49b0f37de37e1f3d with all the discussed changes.